packages/react-ui-components/src/components/SchemaForm/SchemaForm.tsx

```diff
@@ -0,0 +1,111 @@
+import { EActionButtonType, EComponentSize } from '@kelvininc/ui-components';
+import Form, { FormProps, IChangeEvent, utils, withTheme } from '@rjsf/core';
+import classNames from 'classnames';
+import { cloneDeep, isEmpty, isEqualWith } from 'lodash-es';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import { useScroll } from '../../utils';
+import { KvActionButtonText } from '../stencil-generated';
+import styles from './SchemaForm.module.scss';
+import Theme from './Theme';
+
+const ThemedForm = withTheme(Theme);
+const { getSubmitButtonOptions } = utils;
+const SCROLL_OFFSET = 10;
+
+interface SchemaForm<T> extends Form<T> {
+	formElement: HTMLDivElement;
+}
+
+export interface SchemaFormProps<T> extends FormProps<T> {
+	submittedData?: T;
+	allowDiscardChanges?: boolean;
+}
+
+export function KvSchemaForm<T>(props: SchemaFormProps<T>) {
+	const [isValid, setValid] = useState(!props.liveValidate);
+	const [hasChanges, setHasChanges] = useState(!isEqualWith(props.formData, props.submittedData || {}));
+	const [formData, setFormData] = useState(props.formData || ({} as T));
+	const [templateElem, setTemplateElem] = useState<HTMLDivElement | undefined>(undefined);
+
+	const formRef = useRef<SchemaForm<T>>(null);
+	const { scrollTop } = useScroll(templateElem);
+	const isScrolling = useMemo(() => scrollTop - SCROLL_OFFSET > 0, [scrollTop]);
+	const { submitText, norender, props: submitButtonProps } = getSubmitButtonOptions(props.uiSchema || {});
+	const hasFooter = props.allowDiscardChanges || !norender;
+
+	const themedProps = {
+		...props,
+		ref: formRef,
+		onChange: (data: IChangeEvent<T>) => {
+			const hasNewChanges = !isEqualWith(data.formData || {}, props.submittedData || {});
+			setFormData(data.formData);
+			setHasChanges(hasNewChanges);
+			setValid(props.liveValidate ? hasNewChanges && isEmpty(data.errors) : true);
+
+			if (props.onChange != undefined) {
+				props.onChange(data);
+			}
+		},
+		uiSchema: {
+			...props.uiSchema,
+			'ui:submitButtonOptions': {
+				props: { disabled: false },
+				norender: true,
+				submitText: ''
+			}
+		},
+		formData: formData
+	};
+
+	const discardChanges = () => {
+		setFormData(cloneDeep(props.submittedData) as T);
+		setValid(!props.liveValidate);
+		setHasChanges(false);
+	};
+
+	useEffect(() => {
```

Comment:
Can you move this to a custom hook like `useFieldTemplateElement` or something similar?